src/nf/nf_layernorm_submodule.f90

```diff
@@ -0,0 +1,110 @@
+submodule(nf_layernorm_layer) nf_layernorm_layer_submodule
+  implicit none
+contains
+    module function layernorm_layer_cons() &
+    result(res)
+    type(layernorm_layer) :: res
+
+    res % eps = 1e-5
+  end function layernorm_layer_cons
+
+  pure module subroutine forward(self, input)
+    class(layernorm_layer), intent(in out) :: self
+    real, intent(in) :: input(:, :)
+    real, allocatable :: normalized(:, :)
+    integer :: i
+
+    allocate(normalized(self % sequence_length, self % model_dimension))
+
+    ! mu = x - MEAN_last_dim(x)
+    do concurrent(i = 1: self % model_dimension)
+      self % mu(:, i) = input(:, i) - (sum(input, dim=2) / self % model_dimension)
+    end do
+
+    ! square root of variance shifted be eps
+    self % sigma = sqrt((sum(self % mu ** 2, dim=2) / self % model_dimension) + self % eps)
+
+    ! normalize mu by variance by first axis
+    do concurrent(i = 1: self % model_dimension)
+      normalized(:, i) = self % mu(:, i) / self % sigma
+    end do
+
+    ! forward through trainable params gamma and beta
+    do concurrent(i = 1: self % sequence_length)
+      self % output(i, :) = normalized(i, :) * self % gamma + self % beta
+    end do
```

Comment:
It seems that these two loops could be combined, such that the variable `normalized` would not be needed.

Reply:
Yes, I'll fix that

Reply:
Can you explain how would you do that? The only thing I can think of is making a nested loop like this:
```fortran
do concurrent(i = 1: self % sequence_length, j = 1: self % model_dimension)
```
But that'll obviously make it $n \cdot m$ complexity

Reply:
I was indeed thinking to such a nested loop, as it is currently twice the case implicitely. but I might be wrong, and this could be checked at a future stage.

Reply:
In this case I'd definitely use some extra memory to reduce processor complexity to $n + m$